### PR TITLE
storage: create a separate service for each source

### PR DESCRIPTION
**Pull request by benesch**
Create a separate orchestrator service for each source. This means a
process per source when running locally and a pod per source when
running in Kubernetes. This is our storage scalability story for
Materialize Platform.

This will cause some fallout. As written, this creates a process for
every system table, which is hefty.

Fix MaterializeInc/cloud#2708.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Review just the last commit! The first two commits are from #12798 and will be rebased away after that PR merges.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/dataflow-types/src/client/controller/storage.rs
Comment (necaris) [inline src/dataflow-types/src/client/controller/storage.rs:408]: Should this be `assigned.ports["controller"]`, and similarly for below?

File: src/dataflow-types/src/client/controller/storage.rs
Comment (necaris) [inline src/dataflow-types/src/client/controller/storage.rs:420]: No need to block this for a prototype but something we should be careful about, if we care about co-locating `storaged` pods in an AZ for network traffic reasons (ingress / egress, latency)

**Review by necaris — APPROVED** [review]
Looks great so far, per the existing orchestrator semantics in Kubernetes!

**Comment by benesch** [comment]
@petrosagg this is ready for review! I expect all tests, and will rebase after #12798 merges.

**Comment by petrosagg** [comment]
Yikes, that recursive struct mapping introduced quite a bit of boilerplate. I forgot that we rely on the `storaged` instance having the metadata in its local state.

I explored an alternative path that makes storage ingestion descriptions work just like dataflow descriptions where they must carry with them a `source_imports` map with the metadata of any dependent sources. PR here https://github.com/MaterializeInc/materialize/pull/12805 What do you think?

File: src/dataflow-types/src/client/controller/storage.rs
Comment (benesch) [reply src/dataflow-types/src/client/controller/storage.rs:435]: Ah, nope! `assigned` is only in scope on L420-429. The idea is that a service requests a particular port and you may or may not actually get that port. With the process orchestrator, for example, not every `storaged` process can get port 2100. But in Kubernetes since each pod is isolated you can hand out 2100 to every pod.

File: src/dataflow-types/src/client/controller/storage.rs
Comment (benesch) [reply src/dataflow-types/src/client/controller/storage.rs:447]: Yeah, we'll need to think through how we want to allow users to constrain their sources to AZs!

**Pull request merged**
